src/controller/CHIPDeviceController.cpp

```diff
@@ -317,6 +328,88 @@ CHIP_ERROR DeviceController::InitControllerNOCChain(const ControllerInitParams &
     return CHIP_NO_ERROR;
 }
 
+CHIP_ERROR DeviceController::UpdateControllerNOCChain(ControllerInitParams & params)
+{
+    VerifyOrReturnError(mFabricIndex != kUndefinedFabricIndex, CHIP_ERROR_INTERNAL);
+    VerifyOrReturnError(mSystemState != nullptr, CHIP_ERROR_INTERNAL);
+    auto * fabricTable = mSystemState->Fabrics();
+    VerifyOrReturnError(fabricTable->FindFabricWithIndex(mFabricIndex) != nullptr, CHIP_ERROR_INTERNAL);
+
+    constexpr uint32_t chipCertAllocatedLen = kMaxCHIPCertLength;
+    chip::Platform::ScopedMemoryBuffer<uint8_t> originrcacBuf;
+    chip::Platform::ScopedMemoryBuffer<uint8_t> rcacBuf;
+    chip::Platform::ScopedMemoryBuffer<uint8_t> icacBuf;
+    chip::Platform::ScopedMemoryBuffer<uint8_t> nocBuf;
+
+    bool hasExternallyOwnedKeypair                   = false;
+    Crypto::P256Keypair * externalOperationalKeypair = nullptr;
+
+    if (params.operationalKeypair != nullptr)
+    {
+        hasExternallyOwnedKeypair  = params.hasExternallyOwnedOperationalKeypair;
+        externalOperationalKeypair = params.operationalKeypair;
+    }
+
+    ReturnErrorCodeIf(!originrcacBuf.Alloc(chipCertAllocatedLen), CHIP_ERROR_NO_MEMORY);
+    ReturnErrorCodeIf(!rcacBuf.Alloc(chipCertAllocatedLen), CHIP_ERROR_NO_MEMORY);
+    ReturnErrorCodeIf(!icacBuf.Alloc(chipCertAllocatedLen), CHIP_ERROR_NO_MEMORY);
+    ReturnErrorCodeIf(!nocBuf.Alloc(chipCertAllocatedLen), CHIP_ERROR_NO_MEMORY);
+
+    MutableByteSpan originrcacSpan(originrcacBuf.Get(), chipCertAllocatedLen);
+    ReturnErrorOnFailure(fabricTable->FetchRootCert(mFabricIndex, originrcacSpan));
+
+    MutableByteSpan rcacSpan(rcacBuf.Get(), chipCertAllocatedLen);
+    ReturnErrorOnFailure(ConvertX509CertToChipCert(params.controllerRCAC, rcacSpan));
+
+    if (!originrcacSpan.data_equal(rcacSpan))
+    {
+        ChipLogError(Controller, "RCAC is not the same, update NOC chian failed!");
+        return CHIP_ERROR_INVALID_ARGUMENT;
+    }
+
+    MutableByteSpan icacSpan;
+    if (params.controllerICAC.empty())
+    {
+        ChipLogProgress(Controller, "Intermediate CA is not needed");
+    }
+    else
+    {
+        icacSpan = MutableByteSpan(icacBuf.Get(), chipCertAllocatedLen);
+        ReturnErrorOnFailure(ConvertX509CertToChipCert(params.controllerICAC, icacSpan));
+    }
+
+    MutableByteSpan nocSpan = MutableByteSpan(nocBuf.Get(), chipCertAllocatedLen);
+    ReturnErrorOnFailure(ConvertX509CertToChipCert(params.controllerNOC, nocSpan));
+
+    CHIP_ERROR err = CHIP_NO_ERROR;
+
+    auto advertiseOperational =
+        params.enableServerInteractions ? FabricTable::AdvertiseIdentity::Yes : FabricTable::AdvertiseIdentity::No;
+
+    if (externalOperationalKeypair)
+    {
+        err = fabricTable->UpdatePendingFabricWithProvidedOpKey(mFabricIndex, nocSpan, icacSpan, externalOperationalKeypair,
```

Comment:
OK, but now this is duplicating a bunch of fairly fragile code.  Can we please avoid doing that?

Reply:
Add a flag with default value to the function InitControllerNOCChain for updating NOC chain.